deepchem/models/torch_models/dtnn.py

```diff
@@ -0,0 +1,262 @@
+            Number of tasks
+        n_embedding: int, optional
+            Number of features per atom.
+        n_hidden: int, optional
+            Number of features for each molecule after DTNNStep
+        n_distance: int, optional
+            granularity of distance matrix
+            step size will be (distance_max-distance_min)/n_distance
+        distance_min: float, optional
+            minimum distance of atom pairs, default = -1 Angstorm
+        distance_max: float, optional
+            maximum distance of atom pairs, default = 18 Angstorm
+        mode: str
+            Only "regression" is currently supported.
+        dropout: float
+            the dropout probablity to use.
+
+        """
+        super(DTNN, self).__init__()
+        if mode not in ['regression']:
+            raise ValueError("Only 'regression' mode is currently supported")
+        self.n_tasks = n_tasks
+        self.n_embedding = n_embedding
+        self.n_hidden = n_hidden
+        self.n_distance = n_distance
+        self.distance_min = distance_min
+        self.distance_max = distance_max
+        self.step_size = (distance_max - distance_min) / n_distance
+        self.steps = np.array(
+            [distance_min + i * self.step_size for i in range(n_distance)])
+        self.steps = np.expand_dims(self.steps, 0)
+        self.output_activation = output_activation
+        self.mode = mode
+        self.dropout = dropout
+
+        # get DTNNEmbedding
+        self.dtnn_embedding = layers.DTNNEmbedding(n_embedding=self.n_embedding)
+
+        # get DTNNSteps
+        self.dtnn_step = layers.DTNNStep(n_embedding=self.n_embedding,
+                                         n_distance=self.n_distance)
+
+        # get DTNNGather
+        self.dtnn_gather = layers.DTNNGather(
+            n_embedding=self.n_embedding,
+            layer_sizes=[self.n_hidden],
+            n_outputs=self.n_tasks,
+            output_activation=self.output_activation)
+
+        # get Final Linear Layer
+        self.linear = nn.LazyLinear(self.n_tasks)
+
+    def forward(self, inputs: List[torch.Tensor]):
+        """
+        Parameters
+        ----------
+        inputs: List
+            List of Tensors containing.
+            - atom_number
+            - distance
+            - atom_membership
+            - distance_membership_i
+            - distance_membership_j
+
+        Returns
+        -------
+        output: torch.Tensor
+            Predictions of the Molecular Energy.
+
+        """
+        dtnn_embedding = self.dtnn_embedding(inputs[0])
+        if self.dropout > 0.0:
+            dtnn_embedding = nn.Dropout(self.dropout)(dtnn_embedding)
+        dtnn_step = self.dtnn_step(
+            [dtnn_embedding, inputs[1], inputs[3], inputs[4]])
+        if self.dropout > 0.0:
+            dtnn_step = nn.Dropout(self.dropout)(dtnn_step)
+        dtnn_step = self.dtnn_step([dtnn_step, inputs[1], inputs[3], inputs[4]])
+        if self.dropout > 0.0:
```

Comment:
multiple if statements for the same condition ?

Reply:
One more thing that you can do is to validate if self.dropout is greater than 0 in the init function itself instead of forward

Reply:
Can you adress this? Ig we are good to go except this!

Reply:
i think it is ok.
- if we do that if statement check in init function it will not do any benefit.
- and it is assisting the dropout layer. so i think we should keep it as it is.
- it was same in the keras implentation.

Reply:
It is okay you have tried to implement as it is like keras, but the checks are really unnecessary here , that checking the same thing multiple times!

Also, validatingthis dropout and giving error if its less that 0 and greater 1 should be done when the user initilaizes the model itself. In the current design, if I gave an invalid dropout, a part of the function gets run and throws an error, which does not talk about the invalid droput.

I think these changes can make the code a bit clean and debugging easier to future users. And definitely the design is improved from the keras one! Hope you understand!

Reply:
- Implemented the dropout validity check in init.
- Made a common check for dropout > 0.0.

Reply:
I think you may have not got the idea what I was trying to say.
in in the init function, do something like this:
if dropout >=0.0 and dropout <= 1.0:
       self.dropout = dropout
else:
       raise ValueError("dropout value should between 0 and 1")

Then remove the use_dropout attribute all together, and then remove all the checks for self.dropout in forward. If dropout is 0, then dropout would not even occur in forward

Reply:
Implemented those.